File: Constants/TierVendor.lua
require("Classes")
require("GossipEvents")
require("tierVendorGossipEvents")

---@class TierVendor
---@field private entry number
---@field private itemVendor number
---@field private classVendors { [number]: number }
TierVendor = {}
TierVendor.__index = TierVendor

---@param entry number
---@param itemVendor number
---@param classVendors { [number]: number }
function TierVendor.new(entry, itemVendor, classVendors)
    local instance = setmetatable({}, TierVendor)
    instance.entry = entry
    instance.itemVendor = itemVendor
    instance.classVendors = classVendors
    return instance
end


function TierVendor:GetClassVendorEntry(player)
    local playerClass = player:GetClass();
    local vendorId = self.classVendors[playerClass]

    return vendorId
end

function TierVendor:ShowGenericItems(player, creature)
    player:SendListInventory(creature, self.itemVendor)
end

function TierVendor:ShowClassItems(player, creature)
    local classVendorId = self:GetClassVendorEntry(player)
    player:SendListInventory(creature, classVendorId)
end

function TierVendor:GetEntry()
    return self.entry
end

function TierVendor.GossipOnHello(event, player, creature)
    player:GossipClearMenu()

    if (player:GetClass() ~= CLASS_DEATHKNIGHT) then
        player:GossipMenuAddItem(1, "Buy Class Set", 1, SHOW_CLASS_SET)
    end

    player:GossipMenuAddItem(1, "Buy Raid Items", 1, SHOW_GENERIC_ITEMS)

    player:GossipMenuAddItem(1, "Close", 1, CLOSE_GOSSIP)
    player:GossipSendMenu(1, creature, 1)
end

function TierVendor:GossipOnSelectInternal(event, player, creature, sender, intId, code, menuId)
    if (intId == SHOW_CLASS_SET) then
        self:ShowClassItems(player, creature)
    elseif (intId == SHOW_GENERIC_ITEMS) then
        self:ShowGenericItems(player, creature)
    elseif (intId == CLOSE_GOSSIP) then
        player:GossipComplete()
    end
end

-- Necessary to work with the AC Eluna event for registering Gossip, but also keeping the object intact
function TierVendor.GossipOnSelect(tierVendor)
    return function(event, player, creature, sender, intId, code, menuId)
        tierVendor:GossipOnSelectInternal(event, player, creature, sender, intId, code, menuId)
    end
end

File: ts-version/src/TierVendor.ts
import { PlayerClass } from "./Classes";
import { TierVendorGossipEvent } from "./TierVendorGossipEvents";
export class TierVendor {
    constructor(private readonly entry: number, private readonly itemVendor: number, private readonly classVendors: { [key: number]: number }) { }

    GetClassVendorEntry(player: Player) {
        const playerClass = player.GetClass();
        const vendorId = this.classVendors[playerClass];
        return vendorId;
    }

    ShowGenericItems(player: Player, creature: Player | Creature | GameObject | Item) {
        const anyPlayer = player as any; /// Hack until fixed in the definitions
        anyPlayer.SendListInventory(creature, this.itemVendor);
    }

    ShowClassItems(player: Player, creature: Player | Creature | GameObject | Item) {
        const classVendorId = this.GetClassVendorEntry(player);
        const anyPlayer = player as any; /// Hack until fixed in the definitions
        anyPlayer.SendListInventory(creature, classVendorId);
    }

    GetEntry = () => this.entry;

    public static GossipOnHello(player: Player, creature: Creature | GameObject | Item) {
        player.GossipClearMenu();
        if (player.GetClass() != PlayerClass.CLASS_DEATHKNIGHT)
            player.GossipMenuAddItem(1, "Buy Class Set", 1, TierVendorGossipEvent.SHOW_CLASS_SET);

        player.GossipMenuAddItem(1, "Buy Raid Items", 1, TierVendorGossipEvent.SHOW_GENERIC_ITEMS);

        player.GossipMenuAddItem(1, "Close", 1, TierVendorGossipEvent.CLOSE_GOSSIP);
        player.GossipSendMenu(1, creature, 1);

        return true;
    }

    GossipOnSelect(player: Player, creature: Player | Creature | GameObject | Item, intId: number) {
        switch (intId) {
            case TierVendorGossipEvent.SHOW_CLASS_SET:
                this.ShowClassItems(player, creature);
                break;

            case TierVendorGossipEvent.SHOW_GENERIC_ITEMS:
                this.ShowGenericItems(player, creature);
                break;

            case TierVendorGossipEvent.CLOSE_GOSSIP:
                player.GossipComplete();
                break;
        }

        return true;
    }
}